src/map/cashshop.c

```diff
@@ -139,16 +143,314 @@ static int cashshop_read_db_sql( void ){
 	return 0;
 }
 
+#if PACKETVER_SUPPORTS_SALES
+static bool sale_parse_dbrow( char* fields[], int columns, int current ){
+	unsigned short nameid = atoi(fields[0]);
+	int start = atoi(fields[1]), end = atoi(fields[2]), amount = atoi(fields[3]), i;
+	time_t now = time(NULL);
+	struct sale_item_data* sale_item = NULL;
+
+	if( !itemdb_exists(nameid) ){
+		ShowWarning( "sale_parse_dbrow: Invalid ID %hu in line '%d', skipping...\n", nameid, current );
+		return false;
+	}
+
+	ARR_FIND( 0, cash_shop_items[CASHSHOP_TAB_SALE].count, i, cash_shop_items[CASHSHOP_TAB_SALE].item[i]->nameid == nameid );
+
+	if( i == cash_shop_items[CASHSHOP_TAB_SALE].count ){
+		ShowWarning( "sale_parse_dbrow: ID %hu is not registered in the limited tab in line '%d', skipping...\n", nameid, current );
+		return false;
+	}
+
+	// Check if the end is after the start
+	if( start >= end ){
+		ShowWarning( "sale_parse_dbrow: Sale for item %hu was ignored, because the timespan was not correct.\n", nameid );
+		return false;
+	}
+
+	// Check if it is already in the past
+	if( end < now ){
+		ShowWarning( "sale_parse_dbrow: An outdated sale for item %hu was ignored.\n", nameid );
+		return false;
+	}
+
+	// Check if there is already an entry
+	sale_item = sale_find_item(nameid,false);
+
+	if( sale_item == NULL ){
+		RECREATE(sale_items.item, struct sale_item_data *, ++sale_items.count);
+		CREATE(sale_items.item[sale_items.count - 1], struct sale_item_data, 1);
+		sale_item = sale_items.item[sale_items.count - 1];
+	}
+
+	sale_item->nameid = nameid;
+	sale_item->start = start;
+	sale_item->end = end;
+	sale_item->amount = amount;
+	sale_item->timer_start = INVALID_TIMER;
+	sale_item->timer_end = INVALID_TIMER;
+
+	return true;
+}
+
+static void sale_read_db_sql( void ){
+	uint32 lines = 0, count = 0;
+
+	if( SQL_ERROR == Sql_Query( mmysql_handle, "SELECT `nameid`, UNIX_TIMESTAMP(`start`), UNIX_TIMESTAMP(`end`), `amount` FROM `%s` WHERE `end` > now()", sales_table ) ){
+		Sql_ShowDebug(mmysql_handle);
+		return;
+	}
+
+	while( SQL_SUCCESS == Sql_NextRow(mmysql_handle) ){
+		char* str[4];
+		int i;
+
+		lines++;
+
+		for( i = 0; i < 4; i++ ){
+			Sql_GetData( mmysql_handle, i, &str[i], NULL );
+
+			if( str[i] == NULL ){
+				str[i] = "";
+			}
+		}
+
+		if( !sale_parse_dbrow( str, 4, lines ) ){
+			ShowError( "sale_read_db_sql: Cannot process table '%s' at line '%d', skipping...\n", sales_table, lines );
+			continue;
+		}
+
+		count++;
+	}
+
+	Sql_FreeResult(mmysql_handle);
+
+	ShowStatus( "Done reading '"CL_WHITE"%lu"CL_RESET"' entries in '"CL_WHITE"%s"CL_RESET"'.\n", count, sales_table );
+}
+
+static int sale_end_timer( int tid, unsigned int tick, int id, intptr_t data ){
+	struct sale_item_data* sale_item = (struct sale_item_data*)data;
+
+	clif_sale_end( sale_item, NULL, ALL_CLIENT );
+
+	// Remove the timer so the sale end is not sent out again
+	delete_timer( sale_item->timer_end, sale_end_timer );
```

Comment:
I'd delete the timer, before sending to all_client. It just look more secure. If you have tons of client or the broadcast is too slow you could have this rescheduled before the delete. (a strech I know but still)